drivers/serial/uart_nrfx_uart.c

```diff
@@ -209,26 +209,54 @@ static int uart_nrfx_poll_in(struct device *dev, unsigned char *c)
 	return 0;
 }
 
+static void uart_nrfx_isr(void *arg);
+
 /**
  * @brief Output a character in polled mode.
  *
  * @param dev UART device struct
  * @param c Character to send
  */
-static void uart_nrfx_poll_out(struct device *dev,
-					unsigned char c)
+static void uart_nrfx_poll_out(struct device *dev, unsigned char c)
 {
-	/* The UART API dictates that poll_out should wait for the transmitter
-	 * to be empty before sending a character. However, without locking,
-	 * this introduces a rare yet possible race condition if the thread is
-	 * preempted between sending the byte and checking for completion.
-
-	 * Because of this race condition, the while loop has to be placed
-	 * after the write to TXD, and we can't wait for an empty transmitter
-	 * before writing. This is a trade-off between losing a byte once in a
-	 * blue moon against hanging up the whole thread permanently
+	atomic_t *lock;
+#ifdef CONFIG_UART_0_ASYNC
+	while (uart0_cb.tx_buffer) {
+		/* If there is ongoing asynchronous transmission, and we are in
+		 * ISR, then call uart interrupt routine, otherwise
+		 * busy wait until transmission is finished.
+		 */
+		if (k_is_in_isr()) {
+			uart_nrfx_isr((void *) dev);
+		}
+	}
+	/* Use tx_buffer_length as lock, this way uart_nrfx_tx will
+	 * return -EBUSY during poll_out.
 	 */
+	lock = &uart0_cb.tx_buffer_length;
+#else
+	static atomic_val_t poll_out_lock;
 
+	lock = &poll_out_lock;
+#endif
+
+	if (!k_is_in_isr()) {
+		u8_t safety_cnt = 100;
+
+		while (atomic_cas((atomic_t *) lock,
+				  (atomic_val_t) 0,
+				  (atomic_val_t) 1) == false) {
+			/* k_sleep allows other threads to execute and finish
+			 * their transactions.
+			 */
+			k_sleep(1);
+			if (--safety_cnt == 0) {
+				break;
+			}
+		}
+	} else {
+		*lock = 1;
```

Comment:
Lock here is written so that uart_tx returns busy if called from interrupt context.
If another interrupt comes then it is possible that we will lose first character but further calls to poll_out will result in successful transmission